### PR TITLE
DPL Analysis: preliminary work for multiple process() functions

### PR DESCRIPTION
**Pull request by aalkin**
@ktf

### CODE REVIEW

File: Framework/Core/include/Framework/AnalysisTask.h [hunk 1]
```diff
@@ -643,9 +661,11 @@ DataProcessorSpec adaptAnalysisTask(ConfigContext const& ctx, Args&&... args)
   /// make sure options and configurables are set before expression infos are created
   std::apply([&options, &hash](auto&... x) { return (OptionManager<std::decay_t<decltype(x)>>::appendOption(options, x), ...); }, tupledTask);
 
+  auto processTuple = std::make_tuple(&T::process);
```
Comment (ktf) [inline Framework/Core/include/Framework/AnalysisTask.h]: Yes, so basically all we need now is:

```
if constexpr (has_multipleProcessDeclaration<T>) {
   processTuple = T::multipleProcessDeclaration();
} else {
   processTuple = std::make_tuple(&T::process);
}
```

Comment (ktf) [reply Framework/Core/include/Framework/AnalysisTask.h]: actually we can simply pass the processTuple as an argument, defaulting to `{&T::process}` and (almost) be done with the multiple process declaration...

**Comment by aalkin** [comment]
@ktf @jgrosseo this is a working example of multiple process functions in an analysis task. Process functions can have different signatures, only the selected functions will affect what the task subscribes to. The only caveat currently is that filters are applied for each of the process functions, so I'll have to add some caching behavior for selections.

File: Analysis/Tutorials/src/histogramsFullTracks.cxx [hunk 2]
```diff
@@ -18,7 +18,7 @@ using namespace o2;
 using namespace o2::framework;
 using namespace o2::framework::expressions;
 
-struct ATask {
+struct etaAndClsHistograms {
```
Comment (ktf) [inline Analysis/Tutorials/src/histogramsFullTracks.cxx]: Uppercase
```suggestion
struct EtaAndClsHistograms {
```

File: Analysis/Tutorials/src/histogramsFullTracks.cxx [hunk 3]
```diff
@@ -32,6 +32,6 @@ struct ATask {
 WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
 {
   return WorkflowSpec{
-    adaptAnalysisTask<ATask>(cfgc, TaskName{"eta-and-cls-histograms"}),
+    adaptAnalysisTask<etaAndClsHistograms>(cfgc),
```
Comment (ktf) [inline Analysis/Tutorials/src/histogramsFullTracks.cxx]:
```suggestion
    adaptAnalysisTask<EtaAndClsHistograms>(cfgc),
```

**Review by ktf — DISMISSED** [review]
Beautiful. Apart from the capitalisation, I have nothing else to add.

**Comment by jgrosseo** [comment]
Excellent! This was very fast. I am in favour of merging this.

Next, we can look at how to make this more than trivially configurable. 

For instance:
`adaptAnalysisTask<MultipleProcessExample>(cfgc, ProcessesOptional{&MultipleProcessExample::processRec, cfgc.options().get<bool>("doRec"), 
                                                              &MultipleProcessExample::processGen, cfgc.options().get<bool>("doGen"),
                                                              &MultipleProcessExample::processResolution}, cfgc.options().get<bool>("doResolution"))};`

**Comment by aalkin** [comment]
It still needs filter caching, otherwise it would be a performance problem. Related to that is the question whether we also want to have optionally different filters for different processes. 

Configuration like that is not trivial, we'll have to discuss what we can do.

**Comment by ktf** [comment]
why performance problem? doing the filters twice is what would happen right now with two devices, no?

**Comment by aalkin** [comment]
It seems adding proper caching also requires also changing how `Filtered` behaves, currently it always has an internal copy of a selection vector. Since this is already functional I'm marking it ready for review and caching will be the next improvement for this.

**Comment by jgrosseo** [comment]
@aalkin I think the filter caching should anyway go beyond the same process. Our idea is to pull this into a common process if there are several tasks (in independent processes) having the same filter, right?

**Comment by aalkin** [comment]
> @aalkin I think the filter caching should anyway go beyond the same process. Our idea is to pull this into a common process if there are several tasks (in independent processes) having the same filter, right?

At a baseline we need to cache the filter result (row list) in a task, so that it can be supplied to each process() in the task that requires same-ish argument. For example if all process() functions require some sort of join based on `aod::Collisions` we only need to create the row list once and re-use it each time we create the `Filtered` argument for a given process(). I already have an idea on how to implement it, but I need to invent a good static way to determine which process() arguments are equivalent in terms of required filters so I can avoid runtime checks.

**Pull request merged**